### DISCUSSION TITLE
Switch to H.265 as the default video format for target transcoded videofiles in sidecar folder

### DISCUSSION THREAD

**Discussion by kkovaletp** [photoprism/photoprism#3471] [Ideas]
I'm curious why PhotoPrism still transcodes all videos to H.264 format, while there is the successor H.265 with great stability and wide support, but most importantly uses 2-3 times less space for the same quality video. Is there some technical reason stopping the team from switching to the new format? Or that is just kind of legacy and could be optimized?

**Comment by lastzero** [comment]
That's mainly because not everyone has the latest Apple devices and CPU/GPUs for transcoding. Also, as far as I know, there might still be patent issues with H.265, or at least they are more expensive, so some vendors might not support it. Finally, we're committed to merging open pull requests in the coming days and weeks, so that's what we're doing. If you have any helpful information on market share, pros and cons (e.g. software transcoding time vs. H.264), feel free to share! 🤗

**Reply by kkovaletp** [reply]
Thanks for your reply. After that, I googled some more info and tried to transcode a video manually by ffmpeg to H.265, VP9, and AV1 formats. 
As far as I see, both H.26X codecs require some royalty payments. I didn't see the numbers, so couldn't compare which one is cheaper. But I found 2 modern open-source free-for-use codecs: AV1 and VP9, so I've tested them as well.
I found confirmation that H.265 has noticeable less amount of supported devices and platforms compared to H.264. It is strange to me because all major browsers and video players support it and I couldn't imagine a case, where there will be no possibility to play such video (except for really old and outdated devices, which most probably are not even supported by their vendors anymore). Anyway, if the project's mission is to work everywhere, then I'd agree that H.264 is the best choice.
Yes, I use Mac, but my mom uses a Kubuntu laptop with the Intel Core i3, so I understand the performance difference depending on the hardware and platform limitations.
I did my testing using the same source H.264 videos and see that:
- H.265 is the fastest one while producing 2-3 times smaller files with the same quality
- VP9 is ~ 2 times slower than H.265 and the resulting file is bigger: just ~20% smaller than the source
- AV1 is extremely slow (it could process a several mins video during some days), but gives the best file size: up to 7 times smaller

Hope that these findings will help you to make the final decision on the topic, however, my personal opinion is that the product should switch to some modern codec supporting HDR, 4K, smaller file size, and other features of modern smartphones and cameras. Or at least that could be an option on the product Settings page.

**Comment by gigaturbo** [comment]
> I'm curious why PhotoPrism still transcodes all videos to H.264 format, while there is the successor H.265 with great stability and wide support, but most importantly uses 2-3 times less space for the same quality video. Is there some technical reason stopping the team from switching to the new format? Or that is just kind of legacy and could be optimized?

Not all browsers supports H.265 (https://caniuse.com/hevc). From what I understood it will never be supported by Firefox because of the royalties : https://stackoverflow.com/questions/36382568/h-265-hevc-web-browser-support#36390568

AV1 seems to be a good FOSS alternative but is not yet supported by Safari/Edge (or not by default, https://caniuse.com/?search=AV1).

So I think at the moment H.264 is still the best standard codec ...

**Reply by kkovaletp** [reply]
What do you think about VP9?

**Reply by lastzero** [reply]
AV1 is the successor of VP9 and both are not supported by Safari and some other browsers based on this overview (not to mention many hardware encoders/decoders in CPUs and GPUs):
- https://caniuse.com/?search=vp9

See also the comment above:
> VP9 is ~ 2 times slower than H.265 and the resulting file is bigger: just ~20% smaller than the source

**Reply by lastzero** [reply]
That said, you should be able to stream videos you already have in AV1 (and perhaps VP8/VP9) directly with our current release, if your browser supports it. It's just not a good encoding target format if your goal is that it is universally supported (that is the whole point of transcoding).

**Comment by kkovaletp** [comment]
BTW, I found an open alternative to H.265: https://github.com/strukturag/libde265
It is not fully supported by ffmpeg yet, but there is a fork, where it is used for encoding.

Writing this for you to know that there is a royalty-free version...

**Reply by lastzero** [reply]
Finding a library or code is not hard, the problem is support in all browsers and the patents. Money quote:

> On September 29, 2014, [MPEG LA](https://en.wikipedia.org/wiki/MPEG_LA) announced their HEVC license which covers the essential patents from 23 companies.[[48]](https://en.wikipedia.org/wiki/High_Efficiency_Video_Coding#cite_note-MPEGLAHEVCSeptember2014Yahoo-48) The first 100,000 "devices" (which includes software implementations) are royalty free, and after that the fee is $0.20 per device up to an annual cap of $25 million.[[49]](https://en.wikipedia.org/wiki/High_Efficiency_Video_Coding#cite_note-MPEGLAHEVCLicensePDFSeptember2014-49) This is significantly more expensive than the fees on AVC, which were $0.10 per device, with the same 100,000 waiver, and an annual cap of $6.5 million. MPEG LA does not charge any fee on the content itself, something they had attempted when initially licensing AVC, but subsequently dropped when content producers refused to pay it.[[50]](https://en.wikipedia.org/wiki/High_Efficiency_Video_Coding#cite_note-50) The license has been expanded to include the profiles in version 2 of the HEVC standard.

Source: https://en.wikipedia.org/wiki/High_Efficiency_Video_Coding
